dev-tools/mage/manifest/manifest.go

```diff
@@ -251,76 +186,119 @@ func DownloadComponents(ctx context.Context, manifest string, platforms []string
+	if mg.Verbose() {
+		log.Printf(">>>>>>>>>>> Got packagename [%s], looking for exact match", packageName)
+	}
+
+	if exactMatch, ok := project.Packages[packageName]; ok {
+		// We found the exact filename we are looking for
 		if mg.Verbose() {
-			log.Printf(">>>>>>>>>>> Looking for package [%s] of type [%s]", spec.BinaryName, PlatformPackages[platform])
+			log.Printf(">>>>>>>>>>> Found exact match packageName for [%s, %s]: %s", project.Branch, project.CommitHash, exactMatch)
 		}
 
-		var foundIt bool
-		for pkgName := range project.Packages {
-			if strings.HasPrefix(pkgName, spec.BinaryName) {
-				firstSplit := strings.Split(pkgName, spec.BinaryName+"-")
-				if len(firstSplit) < 2 {
-					continue
-				}
+		return &ResolvedPackage{
+			Name: packageName,
+			URLs: []string{exactMatch.URL, exactMatch.ShaURL, exactMatch.AscURL},
+		}, nil
+	}
 
-				secondHalf := firstSplit[1]
-				// Make sure we're finding one w/ the same required package type
-				if strings.Contains(secondHalf, PlatformPackages[platform]) {
-
-					// Split again after the version with the required package string
-					secondSplit := strings.Split(secondHalf, "-"+PlatformPackages[platform])
-					if len(secondSplit) < 2 {
-						continue
-					}
-
-					// The first element after the split should normally be the version
-					pkgVersion := secondSplit[0]
-					if mg.Verbose() {
-						log.Printf(">>>>>>>>>>> Using derived version for package [%s]: %s ", pkgName, pkgVersion)
-					}
-
-					// Create a project/package key with the package, derived version, and required package
-					foundPkgKey := fmt.Sprintf("%s-%s-%s", spec.BinaryName, pkgVersion, PlatformPackages[platform])
-					if mg.Verbose() {
-						log.Printf(">>>>>>>>>>> Looking for project package key: [%s]", foundPkgKey)
-					}
-
-					// Get the package value, if it exists
-					val, ok = project.Packages[foundPkgKey]
-					if !ok {
-						continue
-					}
-
-					if mg.Verbose() {
-						log.Printf(">>>>>>>>>>> Found package key [%s]", foundPkgKey)
-					}
-
-					foundIt = true
-				}
-			}
-		}
+	// If we didn't find it, it may be an Independent Agent Release, where
+	// the opted-in projects will have a patch version one higher than
+	// the rest of the projects, so we "relax" the version constraint
 
-		if !foundIt {
-			return nil, fmt.Errorf("package [%s] not found in project manifest at %s", packageName, project.ExternalArtifactsManifestURL)
-		}
+	// Find the original version in the filename
+	versionIndex := strings.Index(packageName, dependencyVersion)
+	if versionIndex == -1 {
+		return nil, fmt.Errorf("no exact match and filename %q does not seem to contain dependencyVersion %q to try a fallback", packageName, dependencyVersion)
+	}
+
+	// TODO move relaxVersion to the version package so we can rewrite the version like so
+	//parseVersion, _ := version.ParseVersion(dependencyVersion)
+	//parseVersion.GetRelaxedPatchRegexp()
+	relaxedVersion, err := relaxVersion(dependencyVersion)
```

Comment:
nit: wrap the logic of relaxedPackageName in a separate function, I believe this will help highlight that in `ResolveManifestPackage` we first try to do exact matching and then fallback to the relaxed one. Thus if any of the two has any issue we can work our fix in a more scoped footprint 🙂

Reply:
Agreed, `ResolveManifestPackage` has become a bit messy, I will refactor it in the upcoming FIPS packaging PR as there are other modification in the same area of code

Reply:
Refactor done in https://github.com/elastic/elastic-agent/pull/7690